### PR TITLE
Add more cases for passwordless login

### PR DESCRIPTION
**Pull request by Tug**
This PR adds a simple test case for passwordless login on WordPress.com.

It also introduces a new format for defining and using test accounts.
The main idea is to have the ability to define multiple accounts for the same "feature". For instance, here we'd like to test passwordless login, we can then define multiple accounts that have have the `"passworless"` feature:
```json
{
  "accounts": [
	{
	  "email": "passwordless1@gmail.com",
	  "features": [ "passwordless" ]
	},
	{
	  "email": "passwordless2@gmail.com",
	  "features": [ "passwordless" ]
	},
	{
	  "email": "passwordless3@gmail.com",
	  "features": [ "passwordless" ]
	},
	{
	  "email": "passwordless4@gmail.com",
	  "features": [ "passwordless" ]
	}
  ]
}
```

When running the passwordless tests, one of the accounts will be used at a time (most probably) so we should avoid triggering 2 magic link emails for the same account at the same time.

### Testing Instructions
- Apply this patch and use pb/1cc8a for your `config/local-development.json`
- Boot wp-calypso locally
- Run the testsuite: `env BROWSERSIZE=mobile ./node_modules/.bin/mocha specs/wp-log-in-out-spec.js`
- Assert that all the tests pass

### Reviews
- [ ] Product
- [ ] Code

### CODE REVIEW

**Comment by hoverduck** [comment]
I worry that even with the approach you're taking here to routinely clear out the inbox on every query we'll run into timing races when this test is run in parallel.  At a minimum we have a potential for 2 runs at the same time since we do both mobile and desktop screen width, and even beyond that we frequently have multiple builds on CI at the same time.

That's why we opted for the original magic link test to be done on signup, since that's the only way we could guarantee a unique e-mail address that would only receive a single message.  https://github.com/Automattic/wp-e2e-tests/blob/master/specs/wp-signup-spec.js#L65

Is running this test here going to add anything over testing magic links on the signup spec?

**Comment by alisterscott** [comment]
> That's why we opted for the original magic link test to be done on signup, since that's the only way we could guarantee a unique e-mail address that would only receive a single message.

Could we add some meta-data to the magic link email that is sent from WordPress.com that has a GUID or something that we also hide in the Calypso log-in page, so we poll the emails to find the email with the meta-data that we expect which is unique so we don't get clashes even using the same account? Or now I think about that approach do subsequent magic email links immediately invalidate the previous ones?

**Comment by Tug** [comment]
I see, that's a challenging problem!

> Could we add some meta-data to the magic link email that is sent from WordPress.com that has a GUID or something that we also hide in the Calypso log-in page

That would work but I feel it adds quite a lot of complexity. I guess we would need to add it to the email header so it doesn't appear even for text emails...

> Or now I think about that approach do subsequent magic email links immediately invalidate the previous ones?

No they don't invalidate, but they have a validity limit anyway.

How about having multiple test accounts and choose one randomly at the beginning of the test?

**Comment by Tug** [comment]
> Is running this test here going to add anything over testing magic links on the signup spec?

Yes, we'd like to test all login cases, make sure the `redirect_to` parameter is honoured (for jetpack connect for instance), 2fa screen is shown for 2fa users, oauth2 flows etc.

File: lib/flows/login-flow.js [hunk 1]
```diff
@@ -12,28 +12,26 @@ import * as dataHelper from '../data-helper';
...
 
 export default class LoginFlow {
-	constructor( driver, account ) {
+	constructor( driver, accountOrFeatures ) {
 		this.driver = driver;
-
-		if ( account ) {
-			this.account = account;
+		accountOrFeatures = accountOrFeatures || 'defaultUser';
+		if ( typeof accountOrFeatures === 'string' ) {
```
Comment (Tug) [inline lib/flows/login-flow.js]: I don't want to break backward compatibility with existing config so we should make sure any additional code works for both types of account config.

File: lib/data-helper.js [hunk 2]
```diff
@@ -88,6 +88,39 @@ export function getAccountConfig( account ) {
...
+
+	if ( config.has( 'accounts' ) ) {
+		allAccounts = config.get( 'accounts' );
+	} else if ( process.env.ACCOUNT_INFO ) {
+		allAccounts = JSON.parse( process.env.ACCOUNT_INFO );
+	}
+
+	return allAccounts.filter( function( account ) {
+		// return accounts that have all the features requested
+		return difference( account.features || [], features ).length === 0;
```
Comment (yurynix) [inline lib/data-helper.js]: Is that what you wanted here?
```js
_.difference(['passwordless', '2fa'], ['passwordless']); // => [ '2fa' ]
```

shouldn't `features` be a subset of `account.feature` ?

Comment (Tug) [reply lib/data-helper.js]: Nice catch!, thx!

File: lib/data-helper.js [hunk 3]
```diff
@@ -88,6 +88,39 @@ export function getAccountConfig( account ) {
...
+}
+
+export function getRandomAccountWithFeatures( features = [] ) {
+	if ( ! Array.isArray( features ) ) {
+		features = [ features ];
+	}
+
+	const accounts = getAllAccountsWithFeatures( features );
+
+	return accounts[ Math.floor( Math.random() * ( accounts.length -1 ) ) ];
```
Comment (yurynix) [inline lib/data-helper.js]: Shouldn't -1 here, since [Math.random()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random) max is 0.9999999
> The Math.random() function returns a floating-point, pseudo-random number in the range 0 to 1 — that is, from 0 (inclusive) up to but not including 1 (exclusive)

so, for [1,2] you'll get `Math.floor( 0.999999 * 1 )`, never 1

File: lib/email-client.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ const emailWaitMS = config.get( 'emailWaitMS' );
 export default class EmailClient {
 	constructor( mailboxId ) {
 		const apiKey = config.get( 'mailosaurAPIKey' );
-		const Mailosaur = require( 'mailosaur' )( apiKey, 'https://www.mailosaur.com/api/' );
+		const Mailosaur = require( 'mailosaur' )( apiKey, 'https://mailosaur.com/api/' );
```
Comment (yurynix) [inline lib/email-client.js]: Why this change necessary? Wasn't it working before?

File: lib/data-helper.js (hunk 3 above)
Comment (Tug) [reply lib/data-helper.js]: You're right

File: lib/email-client.js (hunk 4 above)
Comment (Tug) [reply lib/email-client.js]: Surprisingly, delete requests returns a 301 redirect and thus fail for `https://www.mailosaur.com/api/` (they were not used in the code until now). Using `https://www.mailosaur.com/api/` it works fine.

File: lib/email-client.js [hunk 5]
```diff
@@ -70,4 +70,19 @@ export default class EmailClient {
 
 		return d.promise;
 	}
+
+	getNewEmailsByRecipient( emailAddress ) {
```
Comment (yurynix) [inline lib/email-client.js]: Super minor, feel free to ignore how about something like that:
```js
getNewEmailsByRecipient( emailAddress ) {
	var self = this;
	return this.pollEmailsByRecipient( emailAddress )
		.then( function( emails ) {
			return self.deleteAllEmail().then( function () { 
				return emails;
			} );
	         } );
} );
```

or we don't care about delete errors:

```js
getNewEmailsByRecipient( emailAddress ) {
	var self = this;
	return this.pollEmailsByRecipient( emailAddress )
		.then( function( emails ) {
		      self.deleteAllEmail();
                      return emails;
	         } );
} );
```

also, why can't we use arrow functions here? we seems to use the ES6 export / import syntax here?

File: specs/wp-log-in-out-spec.js [hunk 6]
```diff
@@ -83,8 +87,63 @@ test.describe( `[${host}] Authentication: (${screenSize}) @parallel @jetpack`, f
...
+				let magicLoginLink, loginFlow, magicLinkEmail, emailClient;
+				test.before( function () {
+					loginFlow = new LoginFlow( driver, [ 'passwordless' ] );
+					emailClient = new EmailClient( get( loginFlow.account, 'mailosaur.inboxId' ) );
+					return loginFlow.login();
+				} );
+
+				test.it( 'Can find the magic link in the email received', function() {
+					return emailClient.pollEmailsByRecipient( loginFlow.account.email ).then( function( emails ) {
+						for ( let email of emails ) {
```
Comment (yurynix) [inline specs/wp-log-in-out-spec.js]: Can we maybe
```js
const selectedEmail = emails.find( email.subject.indexOf( 'WordPress.com' ) > -1 ); // or lodash
magicLinkEmail = selectedEmail;
magicLoginLink = selectedEmail.html.links[0].href;
```

**Comment by yurynix** [comment]
Thank you for working on that @Tug ! 👍 

@hoverduck mentioned:
>At a minimum we have a potential for 2 runs at the same time since we do both mobile and desktop screen width, and even beyond that we frequently have multiple builds on CI at the same time.

From that I understand that multiple test runners might run at the same time on different machines / processes, so the locking and random picking mechanism you've made at `pickRandomAccountWithFeatures()` won't going to solve any collision issues and just makes this unnecessary complicated, but I'm really a noob at this code base, so I might be really mistaken here.

In case I'm not, maybe we should go with a different approach, to assign each testing driver a uuid on the user agent:
```
const selenium = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const uuidv4 = require('uuid/v4');

const opts = new chrome.Options();
opts.addArguments([`user-agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.84 Safari/537.36 e2e_id/${ uuidv4() }"`]);

var driver = new selenium.Builder().
    withCapabilities(opts.toCapabilities()).
    build();
```
This example is for chrome, but similar can be get done [for FF](https://stackoverflow.com/questions/29916054/change-user-agent-for-selenium-driver)

Then we'll get this `e2e_id` on the API and include it in a mail header, which we'll later extract to identify which email we should use.

That's quite complicated, so I would've liked to avoid that if possible, first maybe we should really verify we have really a problem with collisions, maybe start with merging this PR and log to slack channel when test really failed?

What's your thoughts on that @hoverduck @alisterscott ?

**Comment by hoverduck** [comment]
>  I understand that multiple test runners might run at the same time on different machines / processes

That's correct.  So yeah, the account.inUse locking mechanism here won't really come into play.  Ideally I would like to have a truly robust solution in place, like the e2e_id in the e-mail header...but that's also probably unneeded complexity.

Just relying on pure randomness to avoid collisions will _probably_ be okay.  The tests are auto-retried once, so we'd actually have to have a collision two separate times in order to fail the whole suite, so the odds are probably in our favor.
